### DISCUSSION TITLE
Role of DFAs for lexer rules

### DISCUSSION THREAD

**Discussion by DbRRaU6m** [patrickhuber/Earley#108] [Q&A]
Hi, I am currently evaluating Pliant (which I like so far)! However,
I wonder a bit about the lexer rules used in the examples. Apparently they use
a regular expression and d DFA for their definition; could you please give some details about the
role of the DFAs? Perhaps it's obvious for people deeper into parsing, but to me it's not clear at all.

Best regards

**Comment by patrickhuber** [comment]
Hi, thanks for your interest.

A dfa is a deterministic finite automata which is basically a state machine
that has predictable output for a given input.

A very simple state machine would be something like a string of characters
'a' 'b' 'c'. This will recognize or match the string "abc".

You can get more complicated too, where you have loops in the state
machine. For example 'a'+ 'b'+ 'c'+ would recognize strings with one or
more a's followed by one or more b's followed by one or more c's. So "abc"
"aabc" "abbc" "abcc" and on..

The state machine can loop back on itself like a graph of nodes too. The
Wikipedia page does a good job of explains what kinds of dfa are possible
https://en.m.wikipedia.org/wiki/Deterministic_finite_automaton. The biggest
takeaway is that given a character to match, you will always know what
state to move to next. There is no ambiguity.

I decided to go this route instead of using the built in regex library for
dotnet because I could step through the parse one character at a time,
determine why a regular expression didn't parse as well as limit the
capability of the regular expression. For example, dotnet regex have
backtracking, which is inefficient and complicated to implement.

**Reply by DbRRaU6m** [reply]
Well, thanks for your very quick and exhaustive answer,
but it somehow misses the point my question; perhaps it was phrased poorly. I am totally aware of what a DFA is and how it can be used. I was just wondering why _both_ a regex and a DFA are necessary to formulate the lexer rules. Or is the regex just used for illustration while the actual string recognition os done by the DFA?

Best regards

**Comment by patrickhuber** [comment]
Correct. The regex is just a representation of the finite state machine and
the DFA is what does the work. There is a NFA (nondeterministic)
representation in the library as well that serves as an intermediate
representation to turn a regex into a DFA.

**Reply by DbRRaU6m** [reply]
Thanks for your quick reply!

**Reply by DbRRaU6m** [reply]
My apologies to bother you again, but how is it like in PDL then? From the example it looks as if in PDL it is possible to use regular expressions only to define lexer rules, is that correct?

**Reply by patrickhuber** [reply]
Yes, you can use exact strings or regular expressions to define lexer
rules. When the pdl is compiled, regular expressions are ultimately turned into dfa.
